go/chat/unfurl/settings.go

```diff
@@ -0,0 +1,116 @@
+package unfurl
+
+import (
+	"context"
+	"sort"
+
+	"github.com/keybase/client/go/chat/types"
+	"github.com/keybase/client/go/chat/utils"
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/client/go/protocol/chat1"
+)
+
+const settingsModeName = "__unfurl_settings_mode"
+const settingsWhitelistName = "__unfurl_settings_whitelist"
+
+type modeRecord struct {
+	Mode chat1.UnfurlMode
+}
+
+type whitelistRecord struct {
+	Whitelist []string
+}
+
+type Settings struct {
+	utils.DebugLabeler
+
+	storage types.ConversationBackedStorage
+}
+
+func NewSettings(log logger.Logger, storage types.ConversationBackedStorage) *Settings {
+	return &Settings{
+		DebugLabeler: utils.NewDebugLabeler(log, "Settings", false),
+		storage:      storage,
+	}
+}
+
+func (s *Settings) Get(ctx context.Context) (res chat1.UnfurlSettings, err error) {
+	defer s.Trace(ctx, func() error { return err }, "Get")()
+
+	var mr modeRecord
+	found, err := s.storage.Get(ctx, settingsModeName, &mr)
+	if err != nil {
+		return res, err
+	}
+	if !found {
+		s.Debug(ctx, "Get: no mode setting found, using whitelisted")
+		mr.Mode = chat1.UnfurlMode_WHITELISTED
+	}
+
+	var wr whitelistRecord
+	var whitelist []string
+	if found, err = s.storage.Get(ctx, settingsWhitelistName, &wr); err != nil {
+		return res, err
+	}
+	if found {
+		whitelist = wr.Whitelist
+	}
+	return chat1.UnfurlSettings{
+		Mode:      mr.Mode,
+		Whitelist: whitelist,
+	}, nil
+}
+
+func (s *Settings) WhitelistAdd(ctx context.Context, domain string) (err error) {
+	defer s.Trace(ctx, func() error { return err }, "WhitelistAdd(%s)", domain)()
+	var wr whitelistRecord
+	found, err := s.storage.Get(ctx, settingsWhitelistName, &wr)
+	if err != nil {
+		return err
+	}
+	if !found {
+		wr.Whitelist = nil
+	}
+	for _, w := range wr.Whitelist {
```

Comment:
why not make this a map?

Reply:
Yeah probably better, I'll change it.

Reply:
Actually I like it as a list better. The most common operation will be `Get`, and if I change to a map I need to sort it every time (or have some param like should sort). I know the code is a little uglier, but its just contained in this one file.

Reply:
i feel it should only be sorted before it's rendered which is probably the least frequent op.  should there be some `IsValid(domain) bool` method to wrap calling `Get` and handing the mode/whitelist?

Reply:
Changed my mind again, I think map is the way.

Reply:
heheok